### ISSUE TITLE
fast way to make bit-wise copy of a real into a uint?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#12053]
Pulled out of #5054 

@damianmoz was saying:

> Is this technique ever be likely to enable  user to create a bit-wise image of a real(w) into a uint(w), and vica-versa, without any overhead. By this I mean being able to achieve something that the following C does. Otherwise, deep IEEE introspection and manipulation carries a big overhead which would cripple any chance of optimal code.
> ```chapel
> typedef unsigned long _REAL;
> typedef double REAL;
> assert(sizeof(REAL) == sizeof(_REAL));
> typedef union { REAL rawValue;  _REAL bitImage; } _754;
> ...
> REAL x = ...
> ...
> _754 t = { x };
> // at this point t.bitImage will contain the bit-wise image of the floating point number t.rawValue.
> ```

> Note that something like
> ```chapel
> var x = 56.789:real(64)
> var encodingAsPerIEEE754 =  x:void(64):uint(64)
> ...
> var xModified = encodingAsPerIEE754:void(64):real(64);
> ```
> would do the job if the concept of a void or opaque type of a certain number of bytes ever comes into existence. I can do this sort of thing currently with a function call to C but that has totally unacceptable overhead.

**Comment by mppf** [comment]
@damianmoz -

> I can do this sort of thing currently with a function call to C but that has totally unacceptable overhead.

Is there a fundamental reason such a function cannot be inlined? Why couldn't the Chapel compiler simply support a function like `realBitsAsUint` that is inlined by the compiler into a reasonable code sequence? While `void(64)` types might be interesting, this topic doesn't seem to be a good reason to add them vs. simply add a intrinsic-y function that the Chapel compiler knows how to code generate. The intrinsic-y approach is much easier than adding a new built-in type and related special operators.

**Comment by damianmoz** [comment]
I did not realise that the the inlining of a C function was straightforward. Currently, what needs to be done cannot be implemented in Chapel. But if that can be done, that solves the problems.

See my Proposal for that Low Level IEEE 754 Introspection/Manipulation interface in the attached document. It is a first cut (done at airports and in the air) so my apologies for any gaps. The raw document is done in Troff with the MM macros.

[LowLevelIEEE754-ForChapel.pdf](https://github.com/chapel-lang/chapel/files/2771840/LowLevelIEEE754-ForChapel.pdf)

My original idea, since refined in that document, needed this in Chapel
```chapel
//  jam(Target-Type, Source) should really be a primitive implemented as
//
//  inline proc jam(type T, x : S) : T return (x:void(numBits(S))):T;
//
//      targetVar = jam(targetType, sourceVar)
//
//  allows one to copy the data in sourceVar to a variable targetVar
//  of type targetType, supposedly transparently and portably with no
//  conversion of any kind, i.e. just a bit-wise copy of the data

require "jam.h", "jam.c";

extern proc jamu32r32(x : real(32)) : uint(32);
extern proc jamu64r64(x : real(64)) : uint(64);
extern proc jamr32u32(x : uint(32)) : real(32);
extern proc jamr64u64(x : uint(64)) : real(64);

inline proc jam(type T, x : real(32)) where T == uint(32) return jamu32r32(x);
inline proc jam(type T, x : real(64)) where T == uint(64) return jamu64r64(x);
inline proc jam(type T, x : uint(32)) where T == real(32) return jamr32u32(x);
inline proc jam(type T, x : uint(64)) where T == real(64) return jamr64u64(x);
 ```
with the file "jam.h" being
```C
#ifdef CHAPEL_JAM_SOURCE
#define	JAM(N, T, S)	T N(S y) { return *((T *) &y); }
#else
#define	JAM(N, T, S)	extern T N(S y);
#endif

JAM(jamr64u64, double, unsigned long)
JAM(jamu64r64, unsigned long, double)
JAM(jamr32u32, float, unsigned int)
JAM(jamu32r32, unsigned int, float)
```
and "jam.c" being just a few lines
```C
#define	CHAPEL_JAM_SOURCE
#include	"jam.h"
```

**Comment by damianmoz** [comment]
Chapel operates on floating point numbers, those having a form
```chapel
        x = (-1)**n * 2**e * s
```
where n is the negative bit, e is the exponent (stored as a biased value) and s the significand which for the majority of those number satisfies
```chapel
        1 <= |s| < 2
```
It uses these numbers for real(w), imag(w) and complex(w) types. A field or method that Chapel supports on a floating point number is its 'type'. Additionally, complex(w) floating point numbers have two extra read/write fields, 're' and 'im', these allowing access to the real and imaginary components respectively of a complex number.

The following read-only fields are needed for any given Chapel w-bit floating point, real(w) number 'x':
```chapel
        x.ieee754       // returns the IEEE 754 representation of x itself
        x.expbias       // returns the value of the IEEE 754 exponent bias for x.type
        x.sigbits       // returns the number of IEEE 754 significand bits for x.type
```
The first returns the uint(w) representation of the real(w).  As all are interpreted by the compiler, there should be minimal overhead.  The other two are just constants - see appendix below

There also needs to exist an additional routine whose prototype would be
```chapel
        inline proc ieee754(t : uint(?w)) : real(w) return
                //  the floating point number 'y' for which
                //  assert(t == y.ieee754) will yield 'true'
```
Note that
```chapel
        assert(ieee754(x.ieee754) == x); // for all 'x'
```
One can take this further but initial feedback is needed.  This discussion needs work if Chapel will ever treat real(w) as gave a radix other than 2.

**Appendix:**
For completeness we have, for real(w), w = 16, 32, 64, 128 & 256
```chapel
        proc sigBits(param n:integral) param return
            n - (5 + 3 * (n >> 5) - (((n >> 7) + 3 * (n >> 8)) << 1));
        proc sigBits(const n:int(?w)) : int(w) return
            n - (5 + 3 * (n >> 5) - (((n >> 7) + 3 * (n >> 8)) << 1));
        proc expBias(param n:integral) param return
            (1 << (n - sigBits(n) - 1)) - 1;
        proc expBias(const n:int(?w)) : int(w) return
            (1 << (n - sigBits(n) - 1)) - 1;

        assert(x.expbias == expBias(numBits(x.type)); // see below
        assert(x.sigbits == sigBits(numBits(x.type)); // see below

**Examples:**
        param sqrt_2_as_ieee754 = sqrt_2.ieee754;
        // actually what really is needed is the following
        param sqrt2_as_ieee754 = (sqrt_2:real(w)).ieee754;
        // but there are serious problems with param for real(?w)
        // which are yet to be resolved and belong in another issue
```

**Comment by damianmoz** [comment]
Note that the operation of creating the Rvalue x.ieee754, is 1 or 2 assembler instructions on an x86-64, and even less, 0 or 1, on a SPARC, Sorry, no POWER or ARMs here.  And it is even less, depending respectively on whether the real(w) number 'x' is in a register or memory. So the actions being requested here are not rocket science and they provide Chapel with the building blocks for a performant, low-level, IEEE754 interface, something seen to date only in RUST. 

There is an argument that the function mentioned previously, i.e. ieee754, not the read-only field ( or method) should possess a second argument
```chapel
inline proc ieee754(t : uint(w), type T) : T return
   // the floating point number 'y' for which
   // assert(t == y.ieee754) will yield true
   // assert(numBits(T) == w) must hold
```
That second argument gives you the hooks to build in better checking.

The C/C++ equivalent of the ieee754() routine compiles to 2 assembler instructions on X86-64 and 4 on a SPARC. The RISCy SPARC has a 2 instruction overhead to grow and shrink the stack which the CISC x86-64 does not need.

A simpler alternative to the explicit routine is to allow x.ieee754 to be a writable field, i.e.
```chapel
var x : real(w)
...
x.ieee754 = expression of type uint(w); // write-back
...
```
There are some reasons at the back of my mind which says that this looks like a bad thing but I cannot put my finger on it. It may be related to what Chapel will allow you to stuff inside a uint(w). The above statement labelled as **write-back** must have an explicitly uint(w) expression as the Rvalue.

**Comment by damianmoz** [comment]
Just a quick point, ...

The earlier mentioned fields are related to real(w) identifiers, i.e. for a given real(w) _x_, obviously
```chapel
x.ieee754
```
is an attribute of a real(w) expression, the fields
```chapel
x.sigbits
x.expbias
```
are arguably more correctly an attribute of its **type**. Let's not allow arguments about that to slow down any movement on this.

**Comment by mppf** [comment]
@damianmoz 
```
        x.ieee754       // returns the IEEE 754 representation of x itself
        x.expbias       // returns the value of the IEEE 754 exponent bias for x.type
        x.sigbits       // returns the number of IEEE 754 significand bits for x.type
```

> x.sigbits, x.expbias are arguably more correctly an attribute of its **type**

I'm feeling confused about your proposal here.

Here are you imagining that `sigbits` returns 52 for all `real(64)`? And that `expbias` returns 11 in that case?

Or are you imagining that `sigbits` and `expbias` extract those values from the number and then return the bits?

If they are extracting the value from the number, I would expect them to be methods on values of `real(64)` rather than a type method.

(In other words, I'm asking whether you can literally do `real(64).sigbits`. Did you mean something else by "attribute of its type"?)

Lastly, does the sign bit need to be returned in sigbits? So expbias and sigbits would return int? (Or would you rely on another method or just comparisons with 0.0 to extract the sign bit?)

Would you want expbias to return the literal value of the bits in that portion - or the value of the exponent as you'd write it on a chalkboard? (This is relevant since the exponents have an unusual encoding that is not just 2's complement, right?) (And, if they are returning the values you'd write on a board, what would we return for signed zero / subnormals / infinity / NaN, which use special exponent values?)

**Comment by damianmoz** [comment]
To be sure we are speaking almost the same language what do you call the '.type' in
```chapel
	x.type
	x.re or x.im
```
from a complex number 'x'?  Do you prefer
```chapel
	field
	method
	property
```
I hope I am right in assuming that the first implies something that can be both read and written, the second a read-only attribute known at either run or compile time, and the third, a fundamental attribute which most of is the time is known at compile-time. I think that C has member constants and member functions but that choice of language is only sometimes correct for what we are discussing.

Can you give me the Chapel-ese doctrine and I will use that?

Let's get some definitions down because you are mixing things up. My guess is you have too many things on your plate! Tell the boss that he is working you too hard!

On Tue, 19 Feb 2019, Michael Ferguson wrote:

> I'm feeling confused about your proposal here.

That because the underlying model is fundamentally broken so you are not alone. And the worship of backwards compatibility is never too far away either which is a rod for our backs! The floating point model talks about max and min on logarithmic scales while that for integral types works on arithmetic scales! Very confusing!

> ..are you imagining that sigbits returns 52 for all real(64)? And that expbias returns 11 in that case?

Yes. Ask yourself:
```chapel
	How many bits in the significand of a real(64)? - Answer: 53
	How many bits in the significand of a real(64) variable 'x'? : Answer: 53
```
So the answer to the above is yes and I apologize if I previously typed 52 as the number of bits in the significand of a real(64). It is 53

Please note that I am saying significand with a 'D'. There is no 'T' at the end of that word.

While the D language has many flaws, please read
```chapel
	https://dlang.org/spec/property.html
```
This shows that given a real(32) variable 'x', then the following
```chapel
	assert(x.type.sigbits == 24)
	assert(x.type.sigbits == ieee754sigbits(w))
	assert(x.sigbits == 24)
	assert(x.sigbits == ieee754sigbits(w))
```
are perfectly valid and both concepts are needed for productive programming.

> Or are you imagining that sigbits and expbias extract those values from the
> number and then return the bits?

No, They are both is a compile time property, a constant for a given type. They are independent of the value of the floating point number.

I have attached a management summary type document for busy people like yourself. It is incomplete.

> If they are extracting the value from the number, I would expect them to be
> methods on values of real(64) rather than a type method.

They are not extracting any value - anywhere!

I really don't want to get tied up in esoteric arguments. We need to able to say that If
```chapel
	x.type.PROPERTY
```
exists then
```chapel
	x.PROPERTY
```
should exist. Anything else is unproductive and will drive prospective adoptees away. And Chapel cannot afford to do that.

> (In other words, I'm asking whether you can literally do real(64).sigbits.

Yes! Think of it as what C++ calls std:numeric_limits<type>::digits but with a better name. They are only digits if you say they are radix digits. I want a name that really means somthing, i.e. the number of bits in the sig-nificand!

See my initial question.
 
> Lastly, does the sign bit need to be returned in sigbits? So expbias and
> sigbits would return int? (Or would you rely on another method or just
> comparisons with 0.0 to extract the sign bit?)

Let's go back to basics.

We have an IEEE 754 number which is the only type of number I think the entire population should think about. Rejecting totally the outdated notion of floating point model we are burdened with in many languages including both C & C++ which dates from the 1970s, a normal, finite, floating point number
can be expressed as
```chapel
	(-1)^n * 2^e * s     (1)
```
where 's' is the significand 1 <= s < 2, e is the (unbiased) exponent (stored as a biased number e# = e+b where 'b' is some bias and e# > 0, and the 'n' the negative bit. There are 'p' binary digits in the significand. Old fashioned language sometimes calls this a mantissa.  A significand has a leading 1 and then p-1 other digits in what we might call the fraction. There is no such animal as a sign bit. It is a negative bit. There is a sign and it is given by (-1) raised to the value of the negative bit.

There are always 'p' digits in the significand. It never changes for a given type.  For a non-extended IEEE 754 type, i.e. real(16), real(32), real(64), and beyond like real(128), its layout or encoding is defined. If we use 'n' for the bits in the negative bit field, 'e#' for the bits in the (biased) exponent field, and 's' for the bits in the significand field, for a real(32) number, this layout is
```chapel
  [n|e#|e#|e#|e#|e#|e#|e#|e#|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s|s]
```
As you can see, the negative bit has no relationship to the bits in the significand (or mantissa for the historically challenged). There are 23 bits stored in the significand field because the leading bit of the significand is hidden for what IEEE 754 calls binary32 or real(32).
> Would you want expbias to return the literal value of the bits in that
> portion - or the value of the exponent as you'd write it on a chalkboard?
I can say
```chapel
	assert(real(16).expbias == 15);
	assert(real(32).expbias == 127);
	assert(real(64).expbias == 1023);
	assert(real(128).expbias == 16383);
```
The exponent bias never changes for a given type. It think it looks like a literal, compile-time, constant to me. While I do assert that it has no relationship to the value in question, I do want to be able to use the property on the number itself as well as its type. Esoteric arguments should disappear.

The exponent bias does have a relationship to the exponent. From the above, you can see that the biased exponent (not the exponent bias) is stored in the encoding (or representation).  The unbiased exponent which is what we write on the blackboard
```chapel
	e = e# - b
```
Note that the bias is something very different from the exponent. I will not discuss the exponent because that is a function of the value of a number, a run-time feature for anything other than a 'param'.

> (This is relevant since the exponents have an unusual encoding that is not
> just 2's complement, right?)

No. An exponent bias has no relationship to encoding or 2's complement.

> (And, if they are returning the values you'd write on a board, what would we return for signed zero,
> subnormals, infinity / NaN, which use special exponent values?)
See the following
```chapel
	for t in ( +/-0.0, some subnormal number, +/-INFINITY, +/-NAN) do
	{
		for w in (32, 64) do // soon (16, 32, 64, 128)
		{
			type T = real(w);
			const x = t:T;

			assert(x.expbias == x.type.expbias);
			writeln(x.expbias);
		}
	}
```
The only numbers you would see printed are 126 and 1023.

If you tell me where in the compiler you get at the '.type' method of a floating point type, I can probably start on augmenting what it needed.

See the attachment. Sadly we are burdened the nomenclature in floating point limits which says that the largest and smallest things are called max and min. I am never going a win a fight to call them more appropriate names like Huge and Tiny so I have given up.

I will talk about the transmutation stuff another day. Let's get the properties squared away first.
[properties.pdf](https://github.com/chapel-lang/chapel/files/2882760/properties.pdf)

**Comment by damianmoz** [comment]
Sorry - typos. 

... The only numbers you would see printed are 127 and 1023. Fat fingers.

And the genuine cut+paste mistakes from the PDF are fixed in this one:

[properties.pdf](https://github.com/chapel-lang/chapel/files/2882888/properties.pdf)

**Comment by mppf** [comment]
Hi @damianmoz - thanks for the notes. You probably wrote more than you needed to, but I'm definitely clear on what you're asking for re. properties at this point.


> To be sure we are speaking almost the same language what do you call the '.type' in

> ```chapel
> 	x.type
> 	x.re or x.im
> ```

> from a complex number 'x'? Do you prefer

> ```chapel
> 	field
> 	method
> 	property
> ```

> I hope I am right in assuming that the first implies something that can be both read and written, the second a read-only attribute known at either run or compile time, and the third, a fundamental attribute which most of is the time is known at compile-time. I think that C has member constants and member functions but that choice of language is only sometimes correct for what we are discussing.

> Can you give me the Chapel-ese doctrine and I will use that?

`x.type` is actually a special case that has direct compiler support. In general though, in Chapel there are things called parentheses-less methods, which is what `x.re` and `x.im` are for a complex number.

Additionally, in Chapel there can be a method callable on the type. These are called *type methods*.

For example, to take only the simplest thing from your list of properties, I could define them like this in a Chapel module today (and yes there are extra parentheses here for ... reasons...):

``` chapel
// parentheses-less type method
proc type (real(32)).numbits param {
  return 32;
}
// parentheses-less method
proc (real(32)).numbits param {
  return 32;
}

proc type (real(64)).numbits param {
  return 64;
}
proc (real(64)).numbits param {
  return 64;
}
```

And here's some simple code to test them:

``` chapel
var r32:real(32);
var r64:real(32);
writeln(real(32).numbits);
writeln(r32.numbits);
writeln(real(64).numbits);
writeln(r64.numbits);
```

Now, I think it would be reasonable to argue that the `proc type (real(64)).numbits param` version should not be needed, since the compiler should call that method when provided with a call like `r64.numbits`. I personally think that's a reasonable thing to argue for but it's not what is implemented today. Today, if you want a method available both on the type and on values of that type, you have to create two versions of the method. (Of course one version could call the other).
Feel free to create another issue if you'd like to argue in favor of changing that.

In any case I hope that with the above example, you'll find that you are able to create the properties you were hoping for in a module. You could then make this available as a package module, whether or not it were included in the standard library. There might be bug fixes required in the compiler for certain patterns (e.g. where clauses aren't working with parentheses-less type-returning methods, see #11216) but I expect most things will work just fine.

The process of turning the bits in a `real(64)` into a `uint(64)` as you prototyped in jam.h / jam.c is not quite what I'd expect. It's possible to create a robust implementation entirely within Chapel using the `CPtr` module and it's `c_memcpy` function. Using `memcpy` to copy the bits of something into something of a different type is a common trick in C and probably the best way (my opinion) of doing this sort of thing. Unions and the pointer-type-punning you did in the jam.h / jam.c prototype both run into problems with the strict aliasing rules. You might worry that `memcpy` is a complex function and you don't actually want a function call, but compilers commonly optimize this function and know how to (eventually) translate it into bit movement on registers.

Here is an example Chapel source file that does this, with no C interop required:

``` chapel
proc gimmebits(x: real(64)) {
  var src = x;
  var dst: uint(64);
  c_memcpy(c_ptrTo(dst), c_ptrTo(src), c_sizeof(src.type));
  return dst;
}

config const n:real(64) = 16.0;
proc test() {
  var r:real(64) = n;
  var u:uint(64) = gimmebits(r);
  writef("in  = %r\n", r);
  writef("hex = %016xu\n", u);
  writef("bin = %064bu\n", u);
}
test();
```

I hope that with this example you are able to build up a Chapel module that does just what you need here. I'd be happy to explain in more detail any part of this or the other example, just ask.

**Comment by damianmoz** [comment]
@mppf said:

_Today, if you want a method available both on the type and on values of  that type, you have to create two versions of the method. (Of course one  version could call the other). Feel free to create another issue if  you'd like to argue in favor of changing that._

No. I am happy to do the work for both. I will try and do it as a generic function. I hope I do not run into any issues with param and real(32).

> In any case I hope that with the above example, you'll find that you are 
> able to create the properties you were hoping for in a module. You could 
> then make this available as a package module, whether or not it were 
> included in the standard library.

I am happy to create a module and type 'use'.

> There might be bug fixes required in the compiler for certain patterns 
> (e.g. where clauses aren't working with parentheses-less type-returning 
> methods, see #11216) but I expect most things will work just fine.

I assume that I will need to use a where statement. How do you create a parameter'ed method with no parameters to work around that?

Basically, we try and work in real(32) for a variety of reasons. We will  often do development in real(64) and then try to figure out what we need  to do to make things work in real(32). Generally we find that the solution  which works for both real(32) and real(64) is often smarter and far more robust numerically than our original real(64) solution. We need whatever works in real(64) to work for real(32).

> You might worry that memcpy is a complex function and you don't actually 
> want a function call,

You are correct.

> but compilers commonly optimize this function and know how to 
> (eventually) translate it into bit movement on registers.

I wish I had a dollar for every time I have heard that line and it was less than true. That said, a quick test in C says that it optimizes out (today). That has not been the case previously in some non-Chapel projects where  I have been burnt badly on that point.  I have yet to check for the inline case.

> Here is an example Chapel source file that does this, with no C interop required:

I thought I tested that approach about 2 years ago and it gave some woeful performance results. Will try again.
```chapel
proc toieee754(x: real(64))
{
  var src = x;
  var dst: uint(64);
  c_memcpy(c_ptrTo(dst), c_ptrTo(src), c_sizeof(src.type));
  return dst;
}
proc fromieee754(x: uint(64))
{
  var src = x;
  var dst: real(64);
  c_memcpy(c_ptrTo(dst), c_ptrTo(src), c_sizeof(src.type));
  return dst;
}
config const n:real(64) = 16.0;
proc test()
{
  var r:real(64) = n;
  var u:uint(64) = toieee754(r);

  writef("in  = %r\n", r);
  writef("hex = %016xu\n", u);
  writef("bin = %064bu\n", u);
}
test();
```
I have the code 'test.c' inside the --savec=DIR directory called DIR.

How can I compile it to assembler with -O3? There appear to be a gazillion typedefs that are needed and just pulling in "chpl__headerh" is insufficient. I will be very presently surprised if the 19 lines of code maps to the 2 lines of assembler which is what C does with the same memcpy approach.

Sorry, this is a quick response but I hope my reply has the depth of reply it needs to your thoughtful reply.

**Comment by damianmoz** [comment]
Those methods on real(w) - so easy - so cool! Why isn't the whole world using Chapel. Hmmm. Gotta fix those pesky little param real(32)s first to realize the promise of generic programming on all real(w) which for now are just a dream And probably a gazillion other open issues, well over 1000 anyway.

For the methods, I had to resort to
```chapel
proc type (real(?w)).expbias param
{
  return if w == 32 then 127 else 1023;
}
```
I am happy. I will plug my formula in there in the long term.

I tried to see if I can solve the 'where' issue with void parameter list. No. That breaks too.
```chapel
proc type (Real).expbias() param where Real == real(64)
{
  return 1023;
}
```
It tells me that it cannot resolve the base type???

I tried to guess all those typedefs I need to compile the test.c file in the --savec directory containing my toieee754 and fromieee754 routines above. Too many. Cannot find any explanation online. Ideas?

I stuck inline in front of proc above and I am almost convinced that C will do the right thing but it would be good to look at the assembler.

Interestingly, Chapel did the right thing when I said
```chapel
   var one = fromieee754(real(32).expbias:uint(32) << real(32).sigbits);
```
i.e. the compiler figured out that I wanted to do
```chapel
   fromieee754(0x3f800000:uint(32));
```
because that was the number buried in test.c in my --savec directory.

Please keep in the back of your mind that the goal is to have a method **.ieee754** on a real(w) like
```chapel
   assert(x.ieee754() == toieee754(x))
```
This method will  return the IEEE754 encodiing of a const or param, identifier or expression, I tried
```chapel
proc (real(?w)).ieee754
{
   return toieee754(this);
}
```
but it seemed to work. Am I correct?

The compiler will need to eventually have to handle
```chapel
   param sqrt2AsIEEE754 = (sqrt_2:real(w)).ieee754;

   assert(sqrt2AsIEEE754 == 3fb504f3:uint(32)); // if w == 32
```
This way, the compiler is the thing doing that conversion and it is not a run-time overhead. I assume that this **will** need direct compiler support. The current C/C++ compilers do that at compile time with their hands tied behind their back. Going back the other way would be desirable but probably not crucial. Anyway, that is not crucial.

I just tried to pull apart a **param** complex. No dice. I cannot even make a complex(w) work with a compile-time definition. Another one of these floating point numbers that does not handle being treated as a param. The old floating point number type is a bit of a second class citizen in Chapel sometimes. Poor little fundamental type. Oh well, another issue. Where on the list is treatment of even real(32)s as being compile-time definable as a **param**? How can I help to push the fix-priority closer to the top?

I think this issue is almost closed once I check the assembler.  What do you think?

**Comment by mppf** [comment]
> This method will  return the IEEE754 encodiing of a const or param, identifier or expression, I tried
> ```chapel
> proc (real(?w)).ieee754
> {
>    return toieee754(this);
> }
> ```
> but it seemed to work. Am I correct?

Looks good to me.

> I tried to see if I can solve the 'where' issue with void parameter list. No. That breaks too.
>
> ```chapel
> proc type (Real).expbias() param where Real == real(64)
> {
>  return 1023;
> }
> ```

If that were to work today, you'd write

```chapel
proc type (?T).expbias() param where T == real(64)
{ 
  return 1023;
}
```

But it's just not something you can do today.
What you can do is this, which works, if a bit weird:

``` chapel
proc type numeric.expbias() param where this == real(64) || this ==
real(32)
{
  return 1023;
}


writeln(real(32).expbias());
writeln(real(64).expbias());
```


As far as the assembly checking process goes, it can be hard because of extensive inlining. I'll walk you through what I did:

``` chapel
proc gimmebits(x: real(64)) {
  var src = x;
  var dst: uint(64);
  c_memcpy(c_ptrTo(dst), c_ptrTo(src), c_sizeof(src.type));
  return dst;
}

config const n:real(64) = 16.0;
config const iters = 10000;

proc myperftest() {
  var sum:uint = 0;
  for i in 1..iters {
    sum += gimmebits(n);
  }
  return sum;
}
writeln(myperftest());
```

```
chpl program.chpl --llvm --fast --llvm-print-ir myperftest --llvm-print-ir-stage full
```

This prints the LLVM IR after optimization for the function myperftest:
```
; Function Attrs: noinline nounwind readonly
define weak dso_local fastcc i64 @myperftest_chpl() unnamed_addr #0 {
  %1 = load i64, i64* @iters_chpl, align 8, !tbaa !2
  %2 = icmp sgt i64 %1, 0
  %3 = load i64, i64* bitcast (double* @n_chpl to i64*), align 8
  %4 = mul i64 %3, %1
  %5 = select i1 %2, i64 %4, i64 0
  ret i64 %5
}
```

You can see that there is no memcpy call in there and that it translated the loop into a multiply.

What if we want to check the C backend with GCC and x86-64 assembly? It's just harder because we have to convince GCC not to inline everything (because then it can be too hard to find our function). Here I stop the inlining with recursion:

``` chapel
proc gimmebits(x: real(64)) {
  var src = x;
  var dst: uint(64);
  c_memcpy(c_ptrTo(dst), c_ptrTo(src), c_sizeof(src.type));
  return dst;
}

config const thenumber:real(64) = 16.0;
config const niters = 10000;

proc myperftest(n: int, r: real(64)) : uint {
  if n <= 1 then
    return 0;

  var sum:uint = 0;
  sum += myperftest(n-2, r);
  sum += myperftest(n-1, r);
  sum += gimmebits(r);
  return sum;
}

writeln(myperftest(niters, thenumber));
```

```
chpl prog2.chpl  --savec=tmp --fast --ccflags -save-temps 
```

Then I searched in `_main.s` for the function `myperftest`:
```
myperftest_chpl:
.LFB1089:
        .cfi_startproc
.L25:
        cmpq    $1, %rdi
        jle     .L28
        pushq   %r12
        .cfi_def_cfa_offset 16
        .cfi_offset 12, -16
        vmovq   %xmm0, %r12
        pushq   %rbp
        .cfi_def_cfa_offset 24
        .cfi_offset 6, -24
        pushq   %rbx
        .cfi_def_cfa_offset 32
        .cfi_offset 3, -32
        movq    %rdi, %rbx
        leaq    -2(%rdi), %rdi
        call    myperftest_chpl
        leaq    -1(%rbx), %rdi
        vmovq   %r12, %xmm0
        movq    %rax, %rbp
        call    myperftest_chpl
        popq    %rbx
        .cfi_def_cfa_offset 24
        addq    %rbp, %rax
        popq    %rbp
        .cfi_def_cfa_offset 16
        addq    %r12, %rax
        popq    %r12
        .cfi_def_cfa_offset 8
        ret
.L28:
        .cfi_restore 3
        .cfi_restore 6
        .cfi_restore 12
        xorl    %eax, %eax
        ret
        .cfi_endproc
```

and as you can see, there is no `memcpy` call here.

**Comment by damianmoz** [comment]
I tried this:

``` chapel
proc type numeric.expbias param where this == real(64)
{
  return 1023;
}
```
I can live with this for now.

I tried this
```chapel
chpl -O test.chpl  --savec=tmp --fast --ccflags -save-temps 
```
I searched in `_main.s` for memcpy. There were 8 calls. I removed the usage of my method  which called memcpy.  The number of calls never changed. So I guess my the memcpy had no relationship to the memcpy that actually existed inside _main.s. So I am happy that the memcpy routine got optimized out!
 
Using Chapel I can create an interface to manipulate IEEE 754 numbers that is more complete and easier to use than that associated with any other language that I know.  Superior to Fortran or C or C++. I have stayed away from testing for exceptions or clearing them. I create them portably.

There are still some outstanding issues associated with compile-time defined real(w) expressions.
```chapel
        proc (real(64)).epsilon param return 0x1.0p-052:real(64); // works
        proc (real(32)).epsilon param return 0x1.0p-052:real(32); // fails
```
Also
```chapel
        inline proc two param return 2.0:real(64);
        inline proc (real(64)).undertwo param return two - this.epsilon; // works
        proc (real(64)).undertwo param return two - this.epsilon; // fails
```
And as you say
```chapel
        proc type numeric.expbias param where this == real(32) return 127; // works
        proc type (Real).expbias param where Real == real(32) return 127; // fails
```
This is stuff that the C/C++ preprocessor eats for breakfast. We need to fix it. It certainly is stifling my research because run-time evaluation of data in small routines impacts the performance of those same routines.

Where is this stuff in the compiler?  Is the underlying problem known and even better, is the solution known (but the work not allocated)? I can go and look and see it is beyond/within my skill level.

I also believe that a floating point literal should have a higher precision than the largest precision supported by the compiler. Every time it is used in a floating point expression, **const**, **param** or **var** is used, it reduces that precision by the time it is actually hits somewhere in memory or a register.

Looking at my own method
```chapel
        inline proc (real(w)).ieee754 ....
```
it really needs to be handled directly within the compiler because there is still no way to do
```chapel
        param one = (1.0:real(64)).ieee754;
        param two = (2.0:real(w)).ieee754;
        param four = (4.0:Real).ieee754;
```
which is trivial in C/C++. That said, I can live with what I have here. This one is NOT stifling my research.

So, apart from these problems which are separate issues, I think that this issue, #12053, can be closed. There is some overlap of these separate issues with other outstanding issues.

**Comment by mppf** [comment]
@damianmoz - feel free to contribute your module back to the Chapel project once you have something.

> There are still some outstanding issues associated with compile-time defined real(w) expressions.

> ...

> This is stuff that the C/C++ preprocessor eats for breakfast. We need to fix it. It certainly is stifling my research because run-time evaluation of data in small routines impacts the performance of those same routines.

Taking them one at a time:

``` chapel
 proc (real(64)).epsilon param return 0x1.0p-052:real(64); // works
 proc (real(32)).epsilon param return 0x1.0p-052:real(32); // fails
```

Neither of these is working for me. I created #12377 for it. Is this part blocking your progress?

``` chapel
        inline proc two param return 2.0:real(64);
        inline proc (real(64)).undertwo param return two - this.epsilon; // works
        proc (real(64)).undertwo param return two - this.epsilon; // fails
```

I get an error like
```
error: 'param' functions cannot return non-'param' values
```

is that what you are seeing? I think this is waiting on a resolution of #7692 so let's talk about it there.

``` chapel
        proc type numeric.expbias param where this == real(32) return 127; // works
        proc type (Real).expbias param where Real == real(32) return 127; // fails
```

On this one I'm having trouble understanding why
 a) the (Real) version is a "natural" way of writing it (wouldn't it be better as e.g. `proc type real.expbias param ...` ?)
 b) why this is blocking your progress as the `numeric` version works OK.

I'll close this issue but please feel free to open other more specific issues if you feel we don't have one for one of the specific problems here.
